### DISCUSSION TITLE
When `rangeStrategy` is `bump`, renovate fails to update all package.jsons

### DISCUSSION THREAD

**Discussion by pastelsky** [renovatebot/renovate#12101] [Request Help]
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Please select which platform you are using if self-hosting.

github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Latest, run using github bot

### Describe the bug

Whenever `rangeStrategy` is set to `bump`, renovate is supposed to update versions in package.json even if in `yarn.lock` its up to date.

For eg, for package `foo` with latest version `1.1.0`, the following package.json would be updated —

```diff
dependencies: {
- "foo": "^1.0.0"
+  "foo": "^1.1.0" 
}
```

However, there is a bug with renovate when multiple `package.json`s are present in a project — it only updates one package.json, and leaves out the others which might contain `foo`.

This bug only occurs for dependencies that are already up to date in `yarn.lock` and only require a package.json change.

For repro see this repository —

https://github.com/pastelsky/renovate-range-strategy-repro

Here, look at this pull-request that renovate has raised —
https://github.com/pastelsky/renovate-range-strategy-repro/pull/5/files

While it rightly updates `autoprefixer` and `@babel/core` across all package.jsons in the project,
For `redux-thunk` it misses package.jsons in the `packages` folder.

### Expected
It takes all package.jsons into account in the same pull request.

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by pastelsky** [comment]
It did create a separate pull request for the ones it missed —
https://github.com/pastelsky/renovate-range-strategy-repro/pull/3

But this doesn't make sense. Why would it split changes across PRs?
A lot of mon-repos enforce single versions of dependencies across different packages, and this doesn't work for those use cases.

**Reply by rarkins** [reply]
Because you grouped `patch` updates only: https://github.com/pastelsky/renovate-range-strategy-repro/blob/ee79f436220840e2729b4a72c7bfb3de3e8b65f7/renovate.json5#L33-L35

And `redux-thunk` is a minor update, not patch.

**Reply by pastelsky** [reply]
@rarkins That doesn't explain why renovate tried to do a minor version upgrade for the dependency here —
https://github.com/pastelsky/renovate-range-strategy-repro/pull/5/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R7

If it was looking at just patch updates, why would it upgrade it at the root level?

**Reply by rarkins** [reply]
What does the debug message "packageFiles with updates" say about that dependency/upgrade? Maybe the locked version was already at 2.3.x so it treated it as patch.

**Reply by pastelsky** [reply]
>Maybe the locked version was already at 2.3.x so it treated it as patch

Sure yes, the locked version was already at `2.3.x`, but that's the issue. It should either treat the dependency as a patch, and updates all package.jsons, or treats it as a minor, and should not update any package.json.

I don't understand why it's mixing both behaviors here. 

<details>
<summary>This is the output of `packageFiles with updates`</summary>

<pre>
{
  "config": {
    "npm": [
      {
        "packageFile": "package.json",
        "deps": [
          {
            "depType": "dependencies",
            "depName": "redux-thunk",
            "currentValue": "^2.2.0",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "lockedVersion": "2.3.0",
            "depIndex": 0,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/reduxjs/redux-thunk",
            "currentVersion": "2.3.0",
            "isSingleVersion": false,
            "fixedVersion": "2.3.0",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "2.3.0",
                "newValue": "^2.3.0",
                "releaseTimestamp": "2018-05-28T18:44:08.278Z",
                "newMajor": 2,
                "newMinor": 3,
                "updateType": "patch",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/patch-versions-for-packages"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "autoprefixer",
            "currentValue": "^9.8.6",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "lockedVersion": "9.8.6",
            "depIndex": 1,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/postcss/autoprefixer",
            "currentVersion": "9.8.6",
            "isSingleVersion": false,
            "fixedVersion": "9.8.6",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "9.8.8",
                "newValue": "^9.8.8",
                "releaseTimestamp": "2021-10-04T20:25:01.004Z",
                "newMajor": 9,
                "newMinor": 8,
                "updateType": "patch",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/patch-versions-for-packages"
              },
              {
                "bucket": "major",
                "newVersion": "10.3.7",
                "newValue": "^10.3.7",
                "releaseTimestamp": "2021-10-04T20:26:45.254Z",
                "newMajor": 10,
                "newMinor": 3,
                "updateType": "major",
                "isRange": true,
                "branchName": "renovate/autoprefixer-10.x"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/cli",
            "currentValue": "^7.14.8",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "lockedVersion": "7.14.8",
            "depIndex": 2,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-cli",
            "homepage": "https://babel.dev/docs/en/next/babel-cli",
            "currentVersion": "7.14.8",
            "isSingleVersion": false,
            "fixedVersion": "7.14.8",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.7",
                "newValue": "^7.15.7",
                "releaseTimestamp": "2021-09-17T23:06:16.016Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/babel-monorepo"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/core",
            "currentValue": "^7.15.0",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "lockedVersion": "7.15.0",
            "depIndex": 3,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-core",
            "homepage": "https://babel.dev/docs/en/next/babel-core",
            "currentVersion": "7.15.0",
            "isSingleVersion": false,
            "fixedVersion": "7.15.0",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.8",
                "newValue": "^7.15.8",
                "releaseTimestamp": "2021-10-06T20:54:59.139Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "patch",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/patch-versions-for-packages"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/node",
            "currentValue": "^7.14.9",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "lockedVersion": "7.14.9",
            "depIndex": 4,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-node",
            "homepage": "https://babel.dev/docs/en/next/babel-node",
            "currentVersion": "7.14.9",
            "isSingleVersion": false,
            "fixedVersion": "7.14.9",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.8",
                "newValue": "^7.15.8",
                "releaseTimestamp": "2021-10-06T20:54:55.857Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/babel-monorepo"
              }
            ]
          }
        ],
        "packageJsonName": "renovate-repro",
        "packageFileVersion": "1.0.0",
        "packageJsonType": "library",
        "yarnLock": "yarn.lock",
        "managerData": {
          "yarnZeroInstall": false
        },
        "skipInstalls": true,
        "constraints": {},
        "lockFiles": [
          "yarn.lock"
        ]
      },
      {
        "packageFile": "packages/package-a/package.json",
        "deps": [
          {
            "depType": "dependencies",
            "depName": "redux-thunk",
            "currentValue": "^2.2.0",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 0,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/reduxjs/redux-thunk",
            "currentVersion": "2.2.0",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "2.3.0",
                "newValue": "^2.3.0",
                "releaseTimestamp": "2018-05-28T18:44:08.278Z",
                "newMajor": 2,
                "newMinor": 3,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/redux-thunk-2.x"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "autoprefixer",
            "currentValue": "^9.8.6",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 1,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/postcss/autoprefixer",
            "currentVersion": "9.8.6",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "9.8.8",
                "newValue": "^9.8.8",
                "releaseTimestamp": "2021-10-04T20:25:01.004Z",
                "newMajor": 9,
                "newMinor": 8,
                "updateType": "patch",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/patch-versions-for-packages"
              },
              {
                "bucket": "major",
                "newVersion": "10.3.7",
                "newValue": "^10.3.7",
                "releaseTimestamp": "2021-10-04T20:26:45.254Z",
                "newMajor": 10,
                "newMinor": 3,
                "updateType": "major",
                "isRange": true,
                "branchName": "renovate/autoprefixer-10.x"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/cli",
            "currentValue": "^7.14.8",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 2,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-cli",
            "homepage": "https://babel.dev/docs/en/next/babel-cli",
            "currentVersion": "7.14.8",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.7",
                "newValue": "^7.15.7",
                "releaseTimestamp": "2021-09-17T23:06:16.016Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/babel-monorepo"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/core",
            "currentValue": "^7.15.0",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 3,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-core",
            "homepage": "https://babel.dev/docs/en/next/babel-core",
            "currentVersion": "7.15.0",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.8",
                "newValue": "^7.15.8",
                "releaseTimestamp": "2021-10-06T20:54:59.139Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "patch",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/patch-versions-for-packages"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/node",
            "currentValue": "^7.14.9",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 4,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-node",
            "homepage": "https://babel.dev/docs/en/next/babel-node",
            "currentVersion": "7.14.9",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.8",
                "newValue": "^7.15.8",
                "releaseTimestamp": "2021-10-06T20:54:55.857Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/babel-monorepo"
              }
            ]
          }
        ],
        "packageJsonName": "package-a",
        "packageFileVersion": "1.0.0",
        "packageJsonType": "library",
        "managerData": {
          "yarnZeroInstall": false
        },
        "skipInstalls": true,
        "constraints": {}
      },
      {
        "packageFile": "packages/package-b/package.json",
        "deps": [
          {
            "depType": "dependencies",
            "depName": "redux-thunk",
            "currentValue": "^2.2.0",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 0,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/reduxjs/redux-thunk",
            "currentVersion": "2.2.0",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "2.3.0",
                "newValue": "^2.3.0",
                "releaseTimestamp": "2018-05-28T18:44:08.278Z",
                "newMajor": 2,
                "newMinor": 3,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/redux-thunk-2.x"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "autoprefixer",
            "currentValue": "^9.8.6",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 1,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/postcss/autoprefixer",
            "currentVersion": "9.8.6",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "9.8.8",
                "newValue": "^9.8.8",
                "releaseTimestamp": "2021-10-04T20:25:01.004Z",
                "newMajor": 9,
                "newMinor": 8,
                "updateType": "patch",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/patch-versions-for-packages"
              },
              {
                "bucket": "major",
                "newVersion": "10.3.7",
                "newValue": "^10.3.7",
                "releaseTimestamp": "2021-10-04T20:26:45.254Z",
                "newMajor": 10,
                "newMinor": 3,
                "updateType": "major",
                "isRange": true,
                "branchName": "renovate/autoprefixer-10.x"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/cli",
            "currentValue": "^7.14.8",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 2,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-cli",
            "homepage": "https://babel.dev/docs/en/next/babel-cli",
            "currentVersion": "7.14.8",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.7",
                "newValue": "^7.15.7",
                "releaseTimestamp": "2021-09-17T23:06:16.016Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/babel-monorepo"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/core",
            "currentValue": "^7.15.0",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 3,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-core",
            "homepage": "https://babel.dev/docs/en/next/babel-core",
            "currentVersion": "7.15.0",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.8",
                "newValue": "^7.15.8",
                "releaseTimestamp": "2021-10-06T20:54:59.139Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "patch",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/patch-versions-for-packages"
              }
            ]
          },
          {
            "depType": "dependencies",
            "depName": "@babel/node",
            "currentValue": "^7.14.9",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "depIndex": 4,
            "warnings": [],
            "versioning": "npm",
            "sourceUrl": "https://github.com/babel/babel",
            "sourceDirectory": "packages/babel-node",
            "homepage": "https://babel.dev/docs/en/next/babel-node",
            "currentVersion": "7.14.9",
            "isSingleVersion": false,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "7.15.8",
                "newValue": "^7.15.8",
                "releaseTimestamp": "2021-10-06T20:54:55.857Z",
                "newMajor": 7,
                "newMinor": 15,
                "updateType": "minor",
                "isRange": true,
                "isBump": true,
                "branchName": "renovate/babel-monorepo"
              }
            ]
          }
        ],
        "packageJsonName": "package-b",
        "packageFileVersion": "1.0.0",
        "packageJsonType": "library",
        "managerData": {
          "yarnZeroInstall": false
        },
        "skipInstalls": true,
        "constraints": {}
      }
    ]
  }
}
DEBUG: processRepo()
DEBUG: Processing 4 branches: renovate/autoprefixer-10.x, renovate/babel-monorepo, renovate/patch-versions-for-packages, renovate/redux-thunk-2.x
DEBUG: Calculating hourly PRs remaining
DEBUG: Retrieving PR list
DEBUG: Retrieved 6 Pull Requests
DEBUG: currentHourStart=2021-10-11T05:00:00.000+00:00
DEBUG: PR hourly limit remaining: 100
DEBUG: Calculating prConcurrentLimit (30)
DEBUG: getBranchPr(renovate/patch-versions-for-packages)
DEBUG: findPr(renovate/patch-versions-for-packages, undefined, open)
DEBUG: Found PR #5
</pre>

</details>

**Reply by rarkins** [reply]
When a lock file is in place, the determination of whether an update is patch or minor will be done based on the locked version, not what's in the package.json. The locked version is the source of truth for what is actually installed.

Most likely, you shouldn't see this mismatch once you get your repo up to date and go forward.

**Reply by pastelsky** [reply]
Lock file versions being ahead of those in `package.json`s can happen due to various reasons. For example, if you manually run `yarn-deduplicate` on the `lockfile` such that a higher common version is resolved.

Shouldn't that be considered a valid repository state? 

> once you get your repo up to date and go forward

I understand that in this case, I can do a one-time manual sync (and that's what I've done for now), but it shouldn't still be required right?

**Reply by rarkins** [reply]
Nobody's saying it's an invalid repository state. I'm saying that when the locked version doesn't change or changes by a patch then it's treated by Renovate as a patch not minor

**Reply by pastelsky** [reply]
Okay, but isn't it a bug if it fails to update that patch across all package.jsons?

**Reply by rarkins** [reply]
No
